lib/mix/lib/mix/tasks/compile.app.ex

```diff
@@ -202,8 +206,8 @@ defmodule Mix.Tasks.Compile.App do
         :ok
       {:mod, value} ->
         Mix.raise "Application callback module (:mod) should be either [] or {module, start_args}, got: #{inspect value}"
-      _ ->
-        :ok
+      {opt, _value} when is_atom(opt) ->
```

Comment:
Hard to tell from context because of forced push but all keys and env params must be in atoms in an application.